tests/csapi/txnid_test.go

```diff
@@ -22,20 +25,192 @@ func TestTxnInEvent(t *testing.T) {
+	unsignedTxnId := result.Get("unsigned.transaction_id")
+	if !unsignedTxnId.Exists() {
+		t.Fatalf("Event did not have a 'unsigned.transaction_id' on the GET /rooms/%s/event/%s response", roomID, eventID)
+	}
+
+	must.EqualStr(t, unsignedTxnId.Str, txnId, fmt.Sprintf("Event did not have a 'unsigned.transaction_id' on GET /rooms/%s/event/%s response", eventID, roomID))
+}
+
+
+func mustHaveTransactionID(t *testing.T, roomID, eventID, expectedTxnId string) client.SyncCheckOpt {
+	return client.SyncTimelineHas(roomID, func(r gjson.Result) bool {
+		if r.Get("event_id").Str == eventID {
+			unsignedTxnId := r.Get("unsigned.transaction_id")
+			if !unsignedTxnId.Exists() {
+				t.Fatalf("Event %s in room %s should have a 'unsigned.transaction_id', but it did not", eventID, roomID)
+			}
+
+			must.EqualStr(t, unsignedTxnId.Str, expectedTxnId, fmt.Sprintf("Event %s in room %s should have a 'unsigned.transaction_id'", eventID, roomID))
+
+			return true
+		}
+
+		return false
+	})
+}
+
+func mustNotHaveTransactionID(t *testing.T, roomID, eventID string) client.SyncCheckOpt {
+	return client.SyncTimelineHas(roomID, func(r gjson.Result) bool {
+		if r.Get("event_id").Str == eventID {
+			unsignedTxnId := r.Get("unsigned.transaction_id")
+			if unsignedTxnId.Exists() {
+				t.Fatalf("Event %s in room %s should NOT have a 'unsigned.transaction_id', but it did (%s)", eventID, roomID, unsignedTxnId.Str)
+			}
+
+			return true
+		}
+
+		return false
+	})
+}
+
+// TestTxnScopeOnLocalEcho tests that transaction IDs are scoped to the access token, not the device
+// on the sync response
+func TestTxnScopeOnLocalEcho(t *testing.T) {
+	// Conduit scope transaction IDs to the device ID, not the access token.
+	runtime.SkipIf(t, runtime.Conduit)
+
+	deployment := Deploy(t, b.BlueprintCleanHS)
+	defer deployment.Destroy(t)
+
+	deployment.RegisterUser(t, "hs1", "alice", "password", false)
+
+	// Create a first client, which allocates a device ID.
+	c1 := deployment.Client(t, "hs1", "")
+	c1.UserID, c1.AccessToken, c1.DeviceID = c1.LoginUser(t, "alice", "password")
+
+	// Create a room where we can send events.
+	roomID := c1.CreateRoom(t, map[string]interface{}{})
+
+	txnId := "abdefgh"
+	// Let's send an event, and wait for it to appear in the timeline.
+	eventID := c1.SendEventUnsyncedWithTxnID(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "first",
+		},
+	}, txnId)
+
+	// When syncing, we should find the event and it should have a transaction ID on the first client.
+	c1.MustSyncUntil(t, client.SyncReq{}, mustHaveTransactionID(t, roomID, eventID, txnId))
+
+	// Create a second client, inheriting the first device ID.
+	c2 := deployment.Client(t, "hs1", "")
+	c2.UserID, c2.AccessToken, _ = c2.LoginUser(t, "alice", "password", client.WithDeviceID(c1.DeviceID))
+	c2.DeviceID = c1.DeviceID
```

Comment:
Is this something we should be asserting? (I.e. that logging in with a chosen device ID doesn't give you out a new device ID?) Ditto for 155 below.